blazar/plugins/devices/zun_plugin.py

```diff
@@ -117,3 +115,23 @@ def cleanup_device(self, device):
             except Exception as e:
                 LOG.exception('Failed to delete %s: %s.',
                               container.name, str(e))
+
+    def poll_resource_failures(self, devices):
+        failed_devices = []
+        recovered_devices = []
+
+        zun_compute_services = {s.host: s for s in self.zun.services.list()}
+        for device in devices:
```

Comment:
I don't think this should be a nested loop. Instead, you can similarly compute a map between device hostnames and device representations. Then, you can just loop through the list of hostnames one time and compare the Zun versus Blazar state.